cpu: aarch64: add JIT implementation for convolution_f32

m-ymzk opened the pull request:
# Description

This PR adds JIT support of convolution_f32 for AArch64.
It includes JIT implementation for SVE 512.

Related RFC #841
Related PR #850

# Checklist

## Code-change submissions

- [x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally?
  - "benchdnn --conv --batch=inputs/conv/test_conv_ci" is OK. 
  - "test_conv_all" is OK. ~~still running...~~ : (The test_conv_gpu outputted 200 unimplemented.)

- [x] Have you formatted the code using clang-format?

### New features

- [N/A] Have you added relevant tests?
- [N/A] Have you provided motivation for adding a new feature?

There is no need to make a new tests for this PR,
since we can use the existing gtests and benchdnn for this PR.

vpirogov commented:
Hi @m-ymzk, 

Just wanted to let you know that majority of the core team engineers are enjoying the holiday season and time off, so we might be slow with code review until Jan 4th next year.

alokbakshi commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4417 @@
+#include "cpu/aarch64/jit_op_imm_check.hpp"
+
+#define GET_OFF(field) static_cast<int32_t>(offsetof(jit_conv_call_s, field))
+#define KNx_L2_EFFECTIVE_CAPACITY ((512 - 64) * 1024)
```
nit: Is the above `define` relevant here [especially because the prefix `KNx` seems to be specific for `x64` ISA]

m-ymzk replied:
Yes, I certainly don't need to write here. I'll delete it.

alokbakshi replied:
Looks good, thanks!

alokbakshi commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4417 @@
+
+    jcp.nb_ic_blocking = jcp.nb_oc_blocking = 1;
+
+    // Heuristic to optimize code size on KNX
```
nit: Looks like the above comment [and possibly the optimization for `large_code_size` below] is not relevant

m-ymzk replied:
Thx! I'll delete it.

alokbakshi replied:
Looks good, thanks!

alokbakshi commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.hpp:
```diff
@@ -0,0 +1,685 @@
+
+        if (jcp.with_eltwise)
+#ifndef DISABLE_ELTWISE
+            eltwise_injector_ = new jit_uni_eltwise_injector_f32<avx512_common>(
```
Just wondering if we need to substitute `avx512_common ---> sve_512` though as I understand currently `DISABLE_ELTWISE` is set so it does not really matter. Nevertheless it looks bit strange to see `x64` isa here!

m-ymzk replied:
Thank you for pointing it out. It just forgot to fix it.

alokbakshi replied:
Looks good now, thanks!

sotoflom commented on src/cpu/aarch64/cpu_barrier.cpp:
```diff
@@ -108,6 +109,11 @@ struct jit_t : public jit_generator {
 };
 
+void barrier(ctx_t *ctx, int nthr) {
+    static jit_t j; /* XXX: constructed on load ... */
```
```suggestion
    static jit_t j; // constructed during load
```

m-ymzk replied:
Thank you for your suggestion.

sotoflom commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.hpp:
```diff
@@ -0,0 +1,685 @@
+            const jit_conv_conf_t ajcp, const primitive_attr_t &attr)
+        : kernel_(nullptr) {
+        switch (ajcp.oc_block) {
+            case 16:
```
nit: If there are no intentions to support other vector register size variants (!= 512bit), I would suggest removing the templated struct that encapsulates `_jit_aarch64_sve_512_conv_fwd_kernel<16> `.

m-ymzk replied:
Thank you for your suggestion. I will remove these.

sotoflom commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4417 @@
+             * 2) 4FMA Kernel does not support:
+             *  `l_pad > 0 || r_pad > 0`; when `kw > 1`
+             * from incorrect 'get_ow_start' and 'get_ow_end' calculation, so
+             * disable for now. */
```
(I don't think these comments are relevant to this JIT version)

m-ymzk replied:
Thanks for your reviews. I'll delete it.

sotoflom replied:
LGM

sotoflom commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4417 @@
+        return status::unimplemented;
+
+    jcp.ur_w = nstl::min(jcp.ow, regs); // ur_w is min(output width, regs=28)
+    if (jcp.ow > 150) jcp.ur_w = regs;
```
This line isn't necessary.

m-ymzk replied:
I'll delete it.

sotoflom replied:
LGM

sotoflom commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4417 @@
+            }
+        }
+    }
+    if (jcp.simd_w != 16) return status::unimplemented;
```
Above and below there are code paths that work with  `simd_w != 16`, if it's not supported, I suggest they get removed.

m-ymzk replied:
Thanks. I will remove these, too.

sotoflom commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4417 @@
+
+    if (!utils::everyone_is(0, jcp.dilate_d, jcp.dilate_h, jcp.dilate_w)
+            && jcp.ver != ver_fma)
+        return status::unimplemented;
```
I don't see this as required since there are no other versions (`jcp.ver`) for `ver_fma`.

m-ymzk replied:
Thx!
I'll delete this IF-block.

sotoflom commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4417 @@
+        } else if ((pre_offset_input >= 0)
+                && ld1rw_imm_check((i_offset - pre_offset_input) & IMM_MASK12)
+                && !((i_offset - pre_offset_input) & ~IMM_MASK24)) {
+            // The offset from previous access address can be represented by ld1rw imm and a 12-23 bit value
```
I think this line might be > 80 chars.

m-ymzk replied:
I don't know why "clang-format" didn't fix it, but I will fix it by hand.

sotoflom commented on src/cpu/aarch64/jit_aarch64_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4417 @@
+     * JIT-Kernel by unrolling with regards to height instead of width for
+     * the source and filter tensors. The JIT-Kernel also transposes the
+     * strides for the input and filter memory access. */
+    jcp.is_hw_transp = false;
```
Maybe remove all relevant code?

m-ymzk replied:
The setting of this flag was preliminary.
Optimization using this transposition is useful, so I will check it and enable it if there is no problem.

m-ymzk replied:
Sorry, some patterns did not work when this optimization was enabled.
(I overlooked a benchdnn result check using 'test_conv_all'. )
As it stands, it will be invalid again.
Please allow me to leave the associated code for now.

sotoflom commented:
I noticed that most changes belong to a single commit. You might want to separate some items (like changes to gtests) into independent commits (e.g. "gtest: conv: aarch64: add skip for unfused jit").

m-ymzk commented on src/cpu/aarch64/jit_op_imm_check.hpp:
```diff
@@ -0,0 +1,71 @@
+
+// Load a vector register from a memory address generated by a 64-bit scalar base,
+// plus an immediate offset in the range -256 to 255 which is multiplied
+// by the current vector register size in bytes. This instruction is unpredicated.
```
Sorry for the confusion.
This comment is only a note of the ARM-LDR instructions. (c.f. [LDR (vector)](https://developer.arm.com/docs/ddi0596/i/a64-sve-instructions-alphabetic-order/ldr-vector-load-vector-register))
So, "This instruction" was an LDR instruction, and "unpredicated" indicated that PREG would not be used.
I'll delete the sentence and rewrite these comments.

sotoflom replied:
LGM, thanks for the update.

m-ymzk commented on src/cpu/aarch64/jit_op_imm_check.hpp:
```diff
@@ -0,0 +1,71 @@
+    int vlen_shift = cpu_isa_traits<sve_512>::vlen_shift;
+    int shifted_ofs = ofs >> vlen_shift;
+    return ((shifted_ofs) <= LDRMAX) && (shifted_ofs >= LDRMIN)
+            && ((ofs % vlen) == 0);
```
Yes, this function mainly checks the range of "ofs" value. I will rewrite these comments.

sotoflom replied:
LGM, thanks for the update.

m-ymzk commented:
Hi, @sotoflom @alokbakshi and @kawakami-k ,
I made corrections that you pointed out and suggested.
 - Remove unwanted lines
 - Delete Templates : jit_sve_512_conv_fwd_kernel, jit_sve_512_conv_bwd_data_kernel
 - modifying JIT function's and file names : from jit_aarch64_sve_512_... to jit_sve_512_...
 - Change fix for skipping unfused jit in gtest

Please re-review it.

m-ymzk commented:
> - Remove unwanted lines

I restored some lines that should not be removed.

alokbakshi commented on src/cpu/aarch64/cpu_reducer.cpp:
```diff
@@ -0,0 +1,597 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
```
nit: [here and elsewhere] year change in copyright 2020 ---> 2021

m-ymzk replied:
OK, I'll fix it.

sotoflom commented on src/cpu/aarch64/jit_primitive_conf.hpp:
```diff
@@ -0,0 +1,723 @@
+} // namespace impl
+} // namespace dnnl
+
+#endif
```
For this entire file, I would suggest to only include the required fields.

alokbakshi replied:
I second that!

m-ymzk replied:
I'll delete unused code.
I will add it when I need it.

alokbakshi commented on src/cpu/aarch64/jit_primitive_conf.hpp:
```diff
@@ -0,0 +1,723 @@
+    return tag == tag_value ? status::success : status::unimplemented;
+}
+
+struct jit_conv_conf_2x3_wino_t {
```
I was wondering if there is any plan to support Winograd convolution, otherwise we can remove this code.

m-ymzk replied:
I left it because I wanted to support Winograd convolution in the near future, but I will delete it.
I will add it when I need it.

sotoflom commented on src/cpu/aarch64/jit_sve_512_conv_kernel.hpp:
```diff
@@ -0,0 +1,631 @@
+
+    jit_sve_512_conv_fwd_kernel(
+            const jit_conv_conf_t &ajcp, const primitive_attr_t &attr)
+#ifndef DISABLE_ELTWISE
+        : jcp(ajcp)
+        , attr_(attr)
+        , eltwise_injector_(nullptr)
+#else // #ifndef DISABLE_ELTWISE
+        : jcp(ajcp)
+        , attr_(attr)
+#endif // #ifndef DISABLE_ELTWISE
```
```suggestion
        : jcp(ajcp)
        , attr_(attr)
#ifndef DISABLE_ELTWISE
        , eltwise_injector_(nullptr)
#endif // #ifndef DISABLE_ELTWISE
```

m-ymzk replied:
Thank you suggestion. I'll fix it.

sotoflom commented on src/cpu/aarch64/jit_sve_512_conv_kernel.cpp:
```diff
@@ -0,0 +1,4374 @@
+
+    const int full_simd_w = cpu_isa_traits<sve_512>::vlen / typesize;
+    jcp.simd_w = full_simd_w;
+    bool ok_to_try_lower_zmm = true
```
unused, maybe remove?

m-ymzk replied:
Thank you suggestion. I'll delete it.

m-ymzk commented:
Hi, @sotoflom @alokbakshi,
I revised the points you pointed out.
 - removed unwanted definitions especially in src/cpu/aarch64/jit_primitive_conf
 - updated all copyrights in these files from 2020 to 2021
Please re-review it.

sotoflom commented:
LGM, thank you for your contribution!

alokbakshi commented:
Hi @m-ymzk 

It looks good, thanks a lot!

dzarukin commented:
Thanks for your contribution!
